File: api/internal/api/handlers/collaboration.go
package handlers

import (
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/georgi-georgiev/testmesh/internal/storage/models"
	"github.com/georgi-georgiev/testmesh/internal/storage/repository"
	"github.com/google/uuid"
	"go.uber.org/zap"
)

// CollaborationHandler handles collaboration-related requests
type CollaborationHandler struct {
	repo   *repository.CollaborationRepository
	logger *zap.Logger
}

// NewCollaborationHandler creates a new collaboration handler
func NewCollaborationHandler(repo *repository.CollaborationRepository, logger *zap.Logger) *CollaborationHandler {
	return &CollaborationHandler{
		repo:   repo,
		logger: logger,
	}
}

// Presence Handlers

// SetPresenceRequest represents a request to set user presence
type SetPresenceRequest struct {
	UserID       string `json:"user_id" binding:"required"`
	UserName     string `json:"user_name" binding:"required"`
	UserEmail    string `json:"user_email"`
	UserAvatar   string `json:"user_avatar"`
	ResourceType string `json:"resource_type" binding:"required"`
	ResourceID   string `json:"resource_id" binding:"required"`
	Status       string `json:"status"`
	CursorData   string `json:"cursor_data"`
}

// SetPresence handles POST /api/v1/collaboration/presence
func (h *CollaborationHandler) SetPresence(c *gin.Context) {
	var req SetPresenceRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	userID, err := uuid.Parse(req.UserID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID"})
		return
	}

	resourceID, err := uuid.Parse(req.ResourceID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid resource ID"})
		return
	}

	status := req.Status
	if status == "" {
		status = "viewing"
	}

	// Get color index based on user ID
	colorIndex := int(userID[0]) % len(models.PresenceColors)

	presence := &models.UserPresence{
		UserID:       userID,
		UserName:     req.UserName,
		UserEmail:    req.UserEmail,
		UserAvatar:   req.UserAvatar,
		Color:        models.GetPresenceColor(colorIndex),
		ResourceType: req.ResourceType,
		ResourceID:   resourceID,
		Status:       status,
		CursorData:   req.CursorData,
	}

	if err := h.repo.SetPresence(presence); err != nil {
		h.logger.Error("Failed to set presence", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, presence)
}

// RemovePresence handles DELETE /api/v1/collaboration/presence
func (h *CollaborationHandler) RemovePresence(c *gin.Context) {
	userIDStr := c.Query("user_id")
	resourceType := c.Query("resource_type")
	resourceIDStr := c.Query("resource_id")

	userID, err := uuid.Parse(userIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID"})
		return
	}

	resourceID, err := uuid.Parse(resourceIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid resource ID"})
		return
	}

	if err := h.repo.RemovePresence(userID, resourceType, resourceID); err != nil {
		h.logger.Error("Failed to remove presence", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Presence removed"})
}

// GetPresence handles GET /api/v1/collaboration/presence/:resource_type/:resource_id
func (h *CollaborationHandler) GetPresence(c *gin.Context) {
	resourceType := c.Param("resource_type")
	resourceIDStr := c.Param("resource_id")

	resourceID, err := uuid.Parse(resourceIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid resource ID"})
		return
	}

	presences, err := h.repo.GetPresenceForResource(resourceType, resourceID)
	if err != nil {
		h.logger.Error("Failed to get presence", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"presences": presences,
		"count":     len(presences),
	})
}

// Comment Handlers

// CreateCommentRequest represents a request to create a comment
type CreateCommentRequest struct {
	FlowID       string                 `json:"flow_id" binding:"required"`
	StepID       string                 `json:"step_id"`
	ParentID     string                 `json:"parent_id"`
	AuthorID     string                 `json:"author_id" binding:"required"`
	AuthorName   string                 `json:"author_name" binding:"required"`
	AuthorAvatar string                 `json:"author_avatar"`
	Content      string                 `json:"content" binding:"required"`
	Position     map[string]interface{} `json:"position"`
}

// CreateComment handles POST /api/v1/collaboration/comments
func (h *CollaborationHandler) CreateComment(c *gin.Context) {
	var req CreateCommentRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	flowID, err := uuid.Parse(req.FlowID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid flow ID"})
		return
	}

	authorID, err := uuid.Parse(req.AuthorID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid author ID"})
		return
	}

	comment := &models.FlowComment{
		FlowID:       flowID,
		StepID:       req.StepID,
		AuthorID:     authorID,
		AuthorName:   req.AuthorName,
		AuthorAvatar: req.AuthorAvatar,
		Content:      req.Content,
		Position:     req.Position,
	}

	if req.ParentID != "" {
		parentID, err := uuid.Parse(req.ParentID)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid parent ID"})
			return
		}
		comment.ParentID = &parentID
	}

	if err := h.repo.CreateComment(comment); err != nil {
		h.logger.Error("Failed to create comment", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, comment)
}

// GetComment handles GET /api/v1/collaboration/comments/:id
func (h *CollaborationHandler) GetComment(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid comment ID"})
		return
	}

	comment, err := h.repo.GetComment(id)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Comment not found"})
		return
	}

	c.JSON(http.StatusOK, comment)
}

// UpdateCommentRequest represents a request to update a comment
type UpdateCommentRequest struct {
	Content string `json:"content" binding:"required"`
}

// UpdateComment handles PUT /api/v1/collaboration/comments/:id
func (h *CollaborationHandler) UpdateComment(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid comment ID"})
		return
	}

	comment, err := h.repo.GetComment(id)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Comment not found"})
		return
	}

	var req UpdateCommentRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	comment.Content = req.Content
	if err := h.repo.UpdateComment(comment); err != nil {
		h.logger.Error("Failed to update comment", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, comment)
}

// DeleteComment handles DELETE /api/v1/collaboration/comments/:id
func (h *CollaborationHandler) DeleteComment(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid comment ID"})
		return
	}

	if err := h.repo.DeleteComment(id); err != nil {
		h.logger.Error("Failed to delete comment", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Comment deleted"})
}

// ListFlowComments handles GET /api/v1/collaboration/flows/:flow_id/comments
func (h *CollaborationHandler) ListFlowComments(c *gin.Context) {
	flowID, err := uuid.Parse(c.Param("flow_id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid flow ID"})
		return
	}

	includeResolved := c.Query("include_resolved") == "true"

	comments, err := h.repo.ListCommentsForFlow(flowID, includeResolved)
	if err != nil {
		h.logger.Error("Failed to list comments", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"comments": comments,
		"count":    len(comments),
	})
}

// ResolveComment handles POST /api/v1/collaboration/comments/:id/resolve
func (h *CollaborationHandler) ResolveComment(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid comment ID"})
		return
	}

	if err := h.repo.ResolveComment(id); err != nil {
		h.logger.Error("Failed to resolve comment", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Comment resolved"})
}

// UnresolveComment handles POST /api/v1/collaboration/comments/:id/unresolve
func (h *CollaborationHandler) UnresolveComment(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid comment ID"})
		return
	}

	if err := h.repo.UnresolveComment(id); err != nil {
		h.logger.Error("Failed to unresolve comment", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Comment unresolved"})
}

// Activity Feed Handlers

// ListActivity handles GET /api/v1/collaboration/activity
func (h *CollaborationHandler) ListActivity(c *gin.Context) {
	params := repository.ActivityEventParams{}

	if workspaceIDStr := c.Query("workspace_id"); workspaceIDStr != "" {
		if id, err := uuid.Parse(workspaceIDStr); err == nil {
			params.WorkspaceID = id
		}
	}
	if resourceType := c.Query("resource_type"); resourceType != "" {
		params.ResourceType = resourceType
	}
	if resourceIDStr := c.Query("resource_id"); resourceIDStr != "" {
		if id, err := uuid.Parse(resourceIDStr); err == nil {
			params.ResourceID = id
		}
	}
	if actorIDStr := c.Query("actor_id"); actorIDStr != "" {
		if id, err := uuid.Parse(actorIDStr); err == nil {
			params.ActorID = id
		}
	}
	if eventType := c.Query("event_type"); eventType != "" {
		params.EventType = eventType
	}
	if sinceStr := c.Query("since"); sinceStr != "" {
		if since, err := time.Parse(time.RFC3339, sinceStr); err == nil {
			params.Since = since
		}
	}
	if limit, err := strconv.Atoi(c.DefaultQuery("limit", "50")); err == nil {
		params.Limit = limit
	}
	if offset, err := strconv.Atoi(c.DefaultQuery("offset", "0")); err == nil {
		params.Offset = offset
	}

	events, total, err := h.repo.ListActivityEvents(params)
	if err != nil {
		h.logger.Error("Failed to list activity", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"events": events,
		"total":  total,
	})
}

// Flow Version Handlers

// ListFlowVersions handles GET /api/v1/collaboration/flows/:flow_id/versions
func (h *CollaborationHandler) ListFlowVersions(c *gin.Context) {
	flowID, err := uuid.Parse(c.Param("flow_id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid flow ID"})
		return
	}

	limit := 20
	if limitStr := c.Query("limit"); limitStr != "" {
		if l, err := strconv.Atoi(limitStr); err == nil {
			limit = l
		}
	}

	versions, err := h.repo.ListFlowVersions(flowID, limit)
	if err != nil {
		h.logger.Error("Failed to list versions", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"versions": versions,
		"count":    len(versions),
	})
}

// GetFlowVersion handles GET /api/v1/collaboration/flows/:flow_id/versions/:version
func (h *CollaborationHandler) GetFlowVersion(c *gin.Context) {
	flowID, err := uuid.Parse(c.Param("flow_id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid flow ID"})
		return
	}

	version, err := strconv.Atoi(c.Param("version"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid version"})
		return
	}

	v, err := h.repo.GetFlowVersion(flowID, version)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Version not found"})
		return
	}

	c.JSON(http.StatusOK, v)
}

// CompareVersions handles GET /api/v1/collaboration/flows/:flow_id/versions/compare
func (h *CollaborationHandler) CompareVersions(c *gin.Context) {
	flowID, err := uuid.Parse(c.Param("flow_id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid flow ID"})
		return
	}

	v1, err := strconv.Atoi(c.Query("v1"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid v1 parameter"})
		return
	}

	v2, err := strconv.Atoi(c.Query("v2"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid v2 parameter"})
		return
	}

	ver1, ver2, err := h.repo.CompareFlowVersions(flowID, v1, v2)
	if err != nil {
		h.logger.Error("Failed to compare versions", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"version1": ver1,
		"version2": ver2,
	})
}

File: dashboard/lib/api/collaboration.ts
import { apiClient } from './client';

// Types
export interface UserPresence {
  id: string;
  user_id: string;
  user_name: string;
  user_email?: string;
  user_avatar?: string;
  color: string;
  resource_type: string;
  resource_id: string;
  status: 'viewing' | 'editing';
  cursor_data?: string;
  last_active_at: string;
  connected_at: string;
}

export interface FlowComment {
  id: string;
  flow_id: string;
  step_id?: string;
  parent_id?: string;
  author_id: string;
  author_name: string;
  author_avatar?: string;
  content: string;
  resolved: boolean;
  position?: Record<string, any>;
  created_at: string;
  updated_at: string;
  replies?: FlowComment[];
}

export interface ActivityEvent {
  id: string;
  actor_id: string;
  actor_name: string;
  actor_avatar?: string;
  event_type: string;
  resource_type: string;
  resource_id: string;
  resource_name?: string;
  description?: string;
  changes?: Record<string, any>;
  metadata?: Record<string, any>;
  workspace_id?: string;
  created_at: string;
}

export interface FlowVersion {
  id: string;
  flow_id: string;
  version: number;
  content: string;
  author_id: string;
  author_name?: string;
  message?: string;
  description?: string;
  created_at: string;
}

// Presence types
export interface SetPresenceRequest {
  user_id: string;
  user_name: string;
  user_email?: string;
  user_avatar?: string;
  resource_type: string;
  resource_id: string;
  status?: 'viewing' | 'editing';
  cursor_data?: string;
}

// Comment types
export interface CreateCommentRequest {
  flow_id: string;
  step_id?: string;
  parent_id?: string;
  author_id: string;
  author_name: string;
  author_avatar?: string;
  content: string;
  position?: Record<string, any>;
}

// Activity params
export interface ActivityParams {
  workspace_id?: string;
  resource_type?: string;
  resource_id?: string;
  actor_id?: string;
  event_type?: string;
  since?: string;
  limit?: number;
  offset?: number;
}

// Presence API
export async function setPresence(data: SetPresenceRequest): Promise<UserPresence> {
  const response = await apiClient.post('/api/v1/collaboration/presence', data);
  return response.data;
}

export async function removePresence(
  userId: string,
  resourceType: string,
  resourceId: string
): Promise<void> {
  await apiClient.delete('/api/v1/collaboration/presence', {
    params: { user_id: userId, resource_type: resourceType, resource_id: resourceId },
  });
}

export async function getPresence(
  resourceType: string,
  resourceId: string
): Promise<{ presences: UserPresence[]; count: number }> {
  const response = await apiClient.get(
    `/api/v1/collaboration/presence/${resourceType}/${resourceId}`
  );
  return response.data;
}

// Comments API
export async function createComment(data: CreateCommentRequest): Promise<FlowComment> {
  const response = await apiClient.post('/api/v1/collaboration/comments', data);
  return response.data;
}

export async function getComment(id: string): Promise<FlowComment> {
  const response = await apiClient.get(`/api/v1/collaboration/comments/${id}`);
  return response.data;
}

export async function updateComment(id: string, content: string): Promise<FlowComment> {
  const response = await apiClient.put(`/api/v1/collaboration/comments/${id}`, { content });
  return response.data;
}

export async function deleteComment(id: string): Promise<void> {
  await apiClient.delete(`/api/v1/collaboration/comments/${id}`);
}

export async function resolveComment(id: string): Promise<void> {
  await apiClient.post(`/api/v1/collaboration/comments/${id}/resolve`);
}

export async function unresolveComment(id: string): Promise<void> {
  await apiClient.post(`/api/v1/collaboration/comments/${id}/unresolve`);
}

export async function getFlowComments(
  flowId: string,
  includeResolved?: boolean
): Promise<{ comments: FlowComment[]; count: number }> {
  const response = await apiClient.get(`/api/v1/collaboration/flows/${flowId}/comments`, {
    params: { include_resolved: includeResolved },
  });
  return response.data;
}

// Activity API
export async function getActivity(
  params?: ActivityParams
): Promise<{ events: ActivityEvent[]; total: number }> {
  const response = await apiClient.get('/api/v1/collaboration/activity', { params });
  return response.data;
}

// Flow Versions API
export async function getFlowVersions(
  flowId: string,
  limit?: number
): Promise<{ versions: FlowVersion[]; count: number }> {
  const response = await apiClient.get(`/api/v1/collaboration/flows/${flowId}/versions`, {
    params: { limit },
  });
  return response.data;
}

export async function getFlowVersion(flowId: string, version: number): Promise<FlowVersion> {
  const response = await apiClient.get(
    `/api/v1/collaboration/flows/${flowId}/versions/${version}`
  );
  return response.data;
}

export async function compareFlowVersions(
  flowId: string,
  v1: number,
  v2: number
): Promise<{ version1: FlowVersion; version2: FlowVersion }> {
  const response = await apiClient.get(
    `/api/v1/collaboration/flows/${flowId}/versions/compare`,
    { params: { v1, v2 } }
  );
  return response.data;
}

// Event type constants
export const EventTypes = {
  FLOW_CREATED: 'flow.created',
  FLOW_UPDATED: 'flow.updated',
  FLOW_DELETED: 'flow.deleted',
  EXECUTION_STARTED: 'execution.started',
  EXECUTION_COMPLETED: 'execution.completed',
  EXECUTION_FAILED: 'execution.failed',
  COMMENT_ADDED: 'comment.added',
  COMMENT_RESOLVED: 'comment.resolved',
  COLLECTION_CREATED: 'collection.created',
  COLLECTION_UPDATED: 'collection.updated',
  MEMBER_JOINED: 'member.joined',
  MEMBER_LEFT: 'member.left',
} as const;

// Helper to get user-friendly event description
export function getEventDescription(event: ActivityEvent): string {
  const actor = event.actor_name;
  const resource = event.resource_name || event.resource_type;

  switch (event.event_type) {
    case EventTypes.FLOW_CREATED:
      return `${actor} created flow "${resource}"`;
    case EventTypes.FLOW_UPDATED:
      return `${actor} updated flow "${resource}"`;
    case EventTypes.FLOW_DELETED:
      return `${actor} deleted flow "${resource}"`;
    case EventTypes.EXECUTION_STARTED:
      return `${actor} started execution for "${resource}"`;
    case EventTypes.EXECUTION_COMPLETED:
      return `Execution completed for "${resource}"`;
    case EventTypes.EXECUTION_FAILED:
      return `Execution failed for "${resource}"`;
    case EventTypes.COMMENT_ADDED:
      return `${actor} commented on "${resource}"`;
    case EventTypes.COMMENT_RESOLVED:
      return `${actor} resolved a comment on "${resource}"`;
    case EventTypes.COLLECTION_CREATED:
      return `${actor} created collection "${resource}"`;
    case EventTypes.COLLECTION_UPDATED:
      return `${actor} updated collection "${resource}"`;
    case EventTypes.MEMBER_JOINED:
      return `${actor} joined the workspace`;
    case EventTypes.MEMBER_LEFT:
      return `${actor} left the workspace`;
    default:
      return event.description || `${actor} performed ${event.event_type}`;
  }
}

// Helper to get event icon
export function getEventIcon(eventType: string): string {
  switch (eventType) {
    case EventTypes.FLOW_CREATED:
    case EventTypes.COLLECTION_CREATED:
      return '➕';
    case EventTypes.FLOW_UPDATED:
    case EventTypes.COLLECTION_UPDATED:
      return '✏️';
    case EventTypes.FLOW_DELETED:
      return '🗑️';
    case EventTypes.EXECUTION_STARTED:
      return '▶️';
    case EventTypes.EXECUTION_COMPLETED:
      return '✅';
    case EventTypes.EXECUTION_FAILED:
      return '❌';
    case EventTypes.COMMENT_ADDED:
      return '💬';
    case EventTypes.COMMENT_RESOLVED:
      return '✔️';
    case EventTypes.MEMBER_JOINED:
      return '👋';
    case EventTypes.MEMBER_LEFT:
      return '👋';
    default:
      return '📌';
  }
}
